File: SundeeFundee/Domain/BodyLocation.swift
import Foundation

/// Structured body region for injury location mapping.
enum BodyLocation {

    enum Region: String, CaseIterable, Codable {
        case head
        case neck
        case leftShoulder
        case rightShoulder
        case chest
        case upperBack
        case lowerBack
        case leftElbow
        case rightElbow
        case leftWrist
        case rightWrist
        case leftHip
        case rightHip
        case leftKnee
        case rightKnee
        case leftAnkle
        case rightAnkle

        var displayName: String {
            switch self {
            case .head:           return "Head"
            case .neck:           return "Neck"
            case .leftShoulder:   return "Left Shoulder"
            case .rightShoulder:  return "Right Shoulder"
            case .chest:          return "Chest"
            case .upperBack:      return "Upper Back"
            case .lowerBack:      return "Lower Back"
            case .leftElbow:      return "Left Elbow"
            case .rightElbow:     return "Right Elbow"
            case .leftWrist:      return "Left Wrist"
            case .rightWrist:     return "Right Wrist"
            case .leftHip:        return "Left Hip"
            case .rightHip:       return "Right Hip"
            case .leftKnee:       return "Left Knee"
            case .rightKnee:      return "Right Knee"
            case .leftAnkle:      return "Left Ankle"
            case .rightAnkle:     return "Right Ankle"
            }
        }

        /// Maps to the engine's contraindication key.
        var engineKey: String {
            switch self {
            case .head, .neck:                       return "back"
            case .leftShoulder, .rightShoulder:      return "shoulder"
            case .chest:                             return "shoulder"
            case .upperBack:                         return "back"
            case .lowerBack:                         return "back"
            case .leftElbow, .rightElbow:            return "wrist"
            case .leftWrist, .rightWrist:            return "wrist"
            case .leftHip, .rightHip:                return "hip"
            case .leftKnee, .rightKnee:              return "knee"
            case .leftAnkle, .rightAnkle:            return "ankle"
            }
        }
    }

    /// Parses comma-separated raw string into regions.
    static func parseRegions(_ raw: String) -> [Region] {
        raw.split(separator: ",")
            .compactMap { Region(rawValue: String($0).trimmingCharacters(in: .whitespaces)) }
    }

    /// Encodes regions to comma-separated raw string.
    static func encodeRegions(_ regions: [Region]) -> String {
        regions.map(\.rawValue).joined(separator: ",")
    }
}

File: pwa/src/domain/injury/body-location.ts
/**
 * body-location.ts
 * Helper functions for the BodyLocation type.
 * The BodyLocation union type is defined in src/domain/types/index.ts.
 */

import { BodyLocation, BodyLocationEngineKey } from '../types';

/**
 * All valid body location values (matches Swift BodyLocation.Region.allCases).
 */
export const BODY_LOCATIONS: BodyLocation[] = [
  'head',
  'neck',
  'leftShoulder',
  'rightShoulder',
  'chest',
  'upperBack',
  'lowerBack',
  'leftElbow',
  'rightElbow',
  'leftWrist',
  'rightWrist',
  'leftHip',
  'rightHip',
  'leftKnee',
  'rightKnee',
  'leftAnkle',
  'rightAnkle',
];

/**
 * Human-readable display name matching Swift BodyLocation.Region.displayName.
 */
export function bodyLocationDisplayName(location: BodyLocation): string {
  const names: Record<BodyLocation, string> = {
    head: 'Head',
    neck: 'Neck',
    leftShoulder: 'Left Shoulder',
    rightShoulder: 'Right Shoulder',
    chest: 'Chest',
    upperBack: 'Upper Back',
    lowerBack: 'Lower Back',
    leftElbow: 'Left Elbow',
    rightElbow: 'Right Elbow',
    leftWrist: 'Left Wrist',
    rightWrist: 'Right Wrist',
    leftHip: 'Left Hip',
    rightHip: 'Right Hip',
    leftKnee: 'Left Knee',
    rightKnee: 'Right Knee',
    leftAnkle: 'Left Ankle',
    rightAnkle: 'Right Ankle',
  };
  return names[location];
}

/**
 * Maps a BodyLocation to the engine's contraindication key.
 * Matches Swift BodyLocation.Region.engineKey.
 */
export function bodyLocationEngineKeyFromRegion(location: BodyLocation): BodyLocationEngineKey {
  switch (location) {
    case 'head':
    case 'neck':
    case 'upperBack':
    case 'lowerBack':
      return 'back';
    case 'leftShoulder':
    case 'rightShoulder':
    case 'chest':
      return 'shoulder';
    case 'leftElbow':
    case 'rightElbow':
    case 'leftWrist':
    case 'rightWrist':
      return 'wrist';
    case 'leftHip':
    case 'rightHip':
      return 'hip';
    case 'leftKnee':
    case 'rightKnee':
      return 'knee';
    case 'leftAnkle':
    case 'rightAnkle':
      return 'ankle';
  }
}

/**
 * Parses a comma-separated raw string into an array of BodyLocation values.
 * Unknown values are silently dropped (matches Swift compactMap behaviour).
 */
export function parseRegions(raw: string): BodyLocation[] {
  return raw
    .split(',')
    .map(s => s.trim())
    .filter((s): s is BodyLocation => (BODY_LOCATIONS as string[]).includes(s));
}

/**
 * Encodes an array of BodyLocation values to a comma-separated raw string.
 */
export function encodeRegions(regions: BodyLocation[]): string {
  return regions.join(',');
}
